client/sniffer/src/main/java/org/elasticsearch/client/sniff/Sniffer.java

```diff
@@ -51,101 +53,167 @@
     private static final Log logger = LogFactory.getLog(Sniffer.class);
     private static final String SNIFFER_THREAD_NAME = "es_rest_client_sniffer";
 
-    private final Task task;
+    private final HostsSniffer hostsSniffer;
+    private final RestClient restClient;
+    private final long sniffIntervalMillis;
+    private final long sniffAfterFailureDelayMillis;
+    private final Scheduler scheduler;
+    private final AtomicBoolean initialized = new AtomicBoolean(false);
+    private volatile ScheduledTask nextScheduledTask;
 
     Sniffer(RestClient restClient, HostsSniffer hostsSniffer, long sniffInterval, long sniffAfterFailureDelay) {
-        this.task = new Task(hostsSniffer, restClient, sniffInterval, sniffAfterFailureDelay);
+        this(restClient, hostsSniffer, new DefaultScheduler(), sniffInterval, sniffAfterFailureDelay);
+    }
+
+    Sniffer(RestClient restClient, HostsSniffer hostsSniffer, Scheduler scheduler,  long sniffInterval, long sniffAfterFailureDelay) {
+        this.hostsSniffer = hostsSniffer;
+        this.restClient = restClient;
+        this.sniffIntervalMillis = sniffInterval;
+        this.sniffAfterFailureDelayMillis = sniffAfterFailureDelay;
+        this.scheduler = scheduler;
+        /*
+         The first sniffing round is async, so this constructor returns before nextScheduledTask is assigned to a task.
+         The initialized flag is a protection against NPE due to that.
+         */
+        Task task = new Task(sniffIntervalMillis) {
+            @Override
+            public void run() {
+                super.run();
+                initialized.compareAndSet(false, true);
+            }
+        };
+        /*
+         We do not keep track of the returned future as we never intend to cancel the initial sniffing round, we rather
+         prevent any other operation from being executed till the sniffer is properly initialized
+         */
+        scheduler.schedule(task, 0L);
     }
 
     /**
-     * Triggers a new sniffing round and explicitly takes out the failed host provided as argument
+     * Triggers a new immediate sniffing round, which will schedule a new round in sniffAfterFailureDelayMillis
```

Comment:
Maybe "reschedule sniffing to run as soon as possible if it isn't already running".

Reply:
++